internal/controllers/configuration/konglicense_controller.go

```diff
@@ -0,0 +1,381 @@
+				log.V(util.DebugLevel).Info("KongLicense deleted in cluster, delete it in cache")
+				if err := r.LicenseCache.Delete(obj); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// Then pick the effective license in KongLicenses remaining in cache.
+				if err := r.repickLicenseOnDelete(ctx, obj); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	log.V(util.DebugLevel).Info("Reconciling resource", "name", req.Name)
+
+	// clean the object up if it's being deleted
+	if !obj.DeletionTimestamp.IsZero() && time.Now().After(obj.DeletionTimestamp.Time) {
+		log.V(util.DebugLevel).Info("Resource is being deleted, its configuration will be removed", "type", "KongLicense")
+
+		_, objectExistsInCache, err := r.LicenseCache.Get(obj)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if objectExistsInCache {
+			// Delete the object in the cache first.
+			if err := r.LicenseCache.Delete(obj); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Then pick the effective license in KongLicenses remaining in cache.
+			if err := r.repickLicenseOnDelete(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{Requeue: true}, nil // wait until the object is no longer present in the cache
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Add the new/updated KongLicense to cache.
+	if err := r.LicenseCache.Add(obj); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Trigger a compare on stored KongLicenses in cache and pick the newest.
+	chosenLicense := r.pickLicenseInCache()
+	if chosenLicense.Name == obj.Name {
+		log.V(util.DebugLevel).Info("Picked KongLicense being reconciled", "name", obj.Name)
+		err := r.ensureControllerStatusProgrammedCondition(ctx, obj, metav1.ConditionTrue, ConditionReasonPickedAsLatest, "")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		oldChosenLicense := r.getChosenLicense()
+		if oldChosenLicense != nil && oldChosenLicense.Name != chosenLicense.Name {
+			r.Log.V(util.DebugLevel).Info("Originally picked KongLicense replaced", "name", oldChosenLicense.Name)
+			err := r.ensureControllerStatusProgrammedCondition(ctx, oldChosenLicense, metav1.ConditionFalse, ConditionReasonReplacedByNewer, "Replaced by newer created KongLicense")
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		r.setChosenLicense(chosenLicense)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// GetLicense is the interface to get the license in Kong configuration format to use in translator.
+func (r *KongV1Alpha1KongLicenseReconciler) GetLicense() mo.Option[kong.License] {
+	chosenLicense := r.getChosenLicense()
+	if chosenLicense == nil {
+		r.Log.V(util.DebugLevel).Info("No KongLicense available")
+		return mo.None[kong.License]()
+	}
+	r.Log.V(util.DebugLevel).Info("Get license from KongLicense resource", "name", chosenLicense.Name)
+	// TODO: Validate KongLicense on Kong gateway.
```

Comment:
Please consider either removing the TODO comment or supplementing it with a tracking issue link.

Reply:
Done. Added issue #5566.